# Módulo 5 Tarefa 1
## Base de nascidos vivos do DataSUS
O DataSUS disponibiliza diversos arquivos de dados com relação a seus segurados, conforme a [lei da transparência de informações públicas](https://www.sisgov.com/transparencia-acesso-informacao/#:~:text=A%20Lei%20da%20Transpar%C3%AAncia%20(LC,em%20um%20site%20na%20internet.). 

Essas informações podem ser obtidas pela internet [aqui](http://www2.datasus.gov.br/DATASUS/index.php?area=0901&item=1). Como o processo de obtenção desses arquivos foge um pouco do nosso escopo, deixamos o arquivo ```SINASC_RO_2019.csv``` já como vai ser encontrado no DataSUS. O dicionário de dados está no arquivo ```estrutura_sinasc_para_CD.pdf``` (o nome do arquivo tal qual no portal do DataSUS).

### Nosso objetivo
Queremos deixar uma base organizada para podermos estudar a relação entre partos com risco para o bebê e algumas condições como tempo de parto, consultas de pré-natal etc.

#### Preparação da base
1. Carregue a base 'SINASC_RO_2019.csv'. Conte o número de registros e o número de registros não duplicados da base. Dica: você aprendeu um método que remove duplicados, encadeie este método com um outro método que conta o número de linhas. **Há linhas duplicadas?**  

2. Conte o número de valores *missing* por variável.  

3. Ok, no item anterior você deve ter achado pouco prático ler a informação de tantas variáveis, muitas delas nem devem ser interesantes. Então crie uma seleção dessa base somente com as colunas que interessam. São elas:
``` 
['LOCNASC', 'IDADEMAE', 'ESTCIVMAE', 'ESCMAE', 'QTDFILVIVO', 
    'GESTACAO', 'GRAVIDEZ', 'CONSULTAS', 'APGAR5'] 
```
Refaça a contagem de valores *missings*.  

4. Apgar é uma *nota* que o pediatra dá ao bebê quando nasce de acordo com algumas características associadas principalmente à respiração. Apgar 1 e Apgar 5 são as notas 1 e 5 minutos do nascimento. Apgar5 será a nossa variável de interesse principal. Então remova todos os registros com Apgar5 não preenchido. Para esta seleção, conte novamente o número de linhas e o número de *missings*.  

5. observe que as variáveis ```['ESTCIVMAE', 'CONSULTAS']``` possuem o código ```9```, que significa *ignorado*. Vamos assumir que o não preenchido é o mesmo que o código ```9```.<br>
6. Substitua os valores faltantes da quantitativa (```QTDFILVIVO```) por zero.  
7. Das restantes, decida que valore te parece mais adequado (um 'não preenchido' ou um valor 'mais provável' como no item anterior) e preencha. Justifique. Lembre-se de que tratamento de dados é trabalho do cientista, e que estamos tomando decisões a todo o momento - não há necessariamente certo e errado aqui.  
8. O Apgar possui uma classificação indicando se o bebê passou por asfixia:
- Entre 8 e 10 está em uma faixa 'normal'. 
- Entre 6 e 7, significa que o recém-nascido passou por 'asfixia leve'. 
- Entre 4 e 5 significa 'asfixia moderada'.
- Entre 0 e 3 significa 'asfixia severa'.  

Crie uma categorização dessa variável com essa codificação e calcule as frequências dessa categorização.  
<br>
9. Renomeie as variáveis para que fiquem no *snake case*, ou seja, em letras minúsculas, com um *underscore* entre as palávras. Dica: repare que se você não quiser criar um *dataframe* novo, você vai precisar usar a opção ```inplace = True```.

In [1]:
import pandas as pd
import numpy as np
import requests

# 1) Carregue a base 'SINASC_RO_2019.csv'. 

df = pd.read_csv('SINASC_RO_2019.csv')

# Conte o número de registros e o número de registros não duplicados da base. 
# Dica: você aprendeu um método que remove duplicados, encadeie este método com um outro método que conta o número de linhas. 

df2 = df.drop_duplicates()
num_linhas = df.shape[0]
num_linhas2 = df2.shape[0]
print(f"O DataFrame original tem {num_linhas} registros.")
print(f"O DataFrame tratado tem {num_linhas2} registros.")

# Há linhas duplicadas?

if num_linhas == num_linhas2:
    print("Logo, nao há registros duplicados nesta base.")
else:
    print("Logo, há registros duplicados na base.")


O DataFrame original tem 27028 registros.
O DataFrame tratado tem 27028 registros.
Logo, nao há registros duplicados nesta base.


In [2]:
# 2) Conte o número de valores missing por variável.

df.isna().sum()


ORIGEM          0
CODESTAB      115
CODMUNNASC      0
LOCNASC         0
IDADEMAE        0
             ... 
munResUf        0
munResLat       1
munResLon       1
munResAlt       1
munResArea      1
Length: 69, dtype: int64

In [3]:
# 3) Ok, no item anterior você deve ter achado pouco prático ler a informação de tantas variáveis, 
# muitas delas nem devem ser interesantes. Então crie uma seleção dessa base somente com as colunas que interessam.

df[['LOCNASC', 'IDADEMAE', 'ESTCIVMAE', 'ESCMAE', 'QTDFILVIVO', 
    'GESTACAO', 'GRAVIDEZ', 'CONSULTAS', 'APGAR5']].isna().sum()

LOCNASC          0
IDADEMAE         0
ESTCIVMAE      317
ESCMAE         312
QTDFILVIVO    1573
GESTACAO      1232
GRAVIDEZ        79
CONSULTAS        0
APGAR5         103
dtype: int64

In [4]:
# 4) Apgar é uma nota que o pediatra dá ao bebê quando nasce de acordo com algumas características associadas principalmente 
# à respiração. Apgar 1 e Apgar 5 são as notas 1 e 5 minutos do nascimento. Apgar5 será a nossa variável de interesse principal. Então remova todos os registros com Apgar5 não preenchido. 
# Para esta seleção, conte novamente o número de linhas e o número de missings.

num_linhas3 = df[['APGAR1', 'APGAR5']].shape[0]

missing_apgar = df[['APGAR1', 'APGAR5']].isna().sum()

print(f'A seleção possui {num_linhas3} linhas.\nAPGAR 1 tem {missing_apgar[0]} notas faltando.\nAPGAR 5 tem {missing_apgar[1]} notas faltando')

A seleção possui 27028 linhas.
APGAR 1 tem 96 notas faltando.
APGAR 5 tem 103 notas faltando


C:\Users\berna\AppData\Local\Temp\ipykernel_7976\2707725704.py:9: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f'A seleção possui {num_linhas3} linhas.\nAPGAR 1 tem {missing_apgar[0]} notas faltando.\nAPGAR 5 tem {missing_apgar[1]} notas faltando')


In [5]:
# 5) observe que as variáveis ['ESTCIVMAE', 'CONSULTAS'] possuem o código 9, que significa ignorado. 
# Vamos assumir que o não preenchido é o mesmo que o código 9.

nove_est = df['ESTCIVMAE'].value_counts().get(9)
nove_con = df['CONSULTAS'].value_counts().get(9)

print(f'O valor 9 aparece {nove_est} vezes em ESTCIVMAE e {nove_con} em CONSULTAS.')

O valor 9 aparece 145 vezes em ESTCIVMAE e 51 em CONSULTAS.


In [6]:
# 6) Substitua os valores faltantes da quantitativa (QTDFILVIVO) por zero.

df['QTDFILVIVO'].fillna(0)

0        0.0
1        1.0
2        2.0
3        0.0
4        1.0
        ... 
27023    1.0
27024    0.0
27025    0.0
27026    1.0
27027    1.0
Name: QTDFILVIVO, Length: 27028, dtype: float64

In [7]:
# 7) Das restantes, decida que valore te parece mais adequado (um 'não preenchido' ou um valor 'mais provável' como no item 
# anterior) e preencha. Justifique. Lembre-se de que tratamento de dados é trabalho do cientista, e que estamos tomando 
# decisões a todo o momento - não há necessariamente certo e errado aqui.

df['ESCMAE'].fillna('Não Preenchido')
df['GESTACAO'].fillna('Não Preenchido')
df['GRAVIDEZ'].fillna('Única')
df['APGAR5'].fillna('9')
df['ESTCIVMAE'].fillna('9')

0        5.0
1        2.0
2        9.0
3        5.0
4        2.0
        ... 
27023    2.0
27024    1.0
27025    2.0
27026    2.0
27027    5.0
Name: ESTCIVMAE, Length: 27028, dtype: object

JUSTIFICATIVAS DO ALUNO:
- Como ESCMAE representa a escolaridade da mãe, nesse caso, não acho que faz sentido estimar um valor para substituir. Neste caso substitui por 'Não Preenchido".
- Como GESTACAO possui uma frequência alta de valores faltantes (1232 unidades), poderia alterar muito o que análises de dados indicaria. Sendo assim, também preferi substituir por 'Não Preenchido'.
- Como GRAVIDEZ possui 'Única' como valor altamente dominante e gravidez duplas ou triplas são muito mais raras, acredito que o melhor tratamento nesse caso seja alterar os valores faltantes para 'Única'.
- Como APGAR5 tem um valor muito baixo de dados faltantes (apenas 103)' e um valor muito dominante (valor 9.0 com 16076 contagens), acho que faz sentido alocar os dados faltantes no dado mais provável.
- No caso de ESTCIVMAE, como '9' significa 'ignorado', aloquei os valores faltantes nele.

In [8]:
# 8) O Apgar possui uma classificação indicando se o bebê passou por asfixia:
# Entre 8 e 10 está em uma faixa 'normal'.
# Entre 6 e 7, significa que o recém-nascido passou por 'asfixia leve'.
# Entre 4 e 5 significa 'asfixia moderada'.
# Entre 0 e 3 significa 'asfixia severa'.

#df.loc[df['APGAR5'] <= 3, 'asfixia'] = 'asfixia severa'
#df.loc[(df['APGAR5'] >= 4) or (df['APGAR5'] <= 5), 'asfixia'] = 'asfixia moderada'
#df.loc[(df['APGAR5'] >= 6) or (df['APGAR5'] <= 7), 'asfixia'] = 'asfixia leve'
#df.loc[(df['APGAR5'] >= 8), 'asfixia'] = 'normal'

df['asfixia'] = np.where(
    df['APGAR5'] <=3 , 'asfixia severa',
    np.where(df['APGAR5'] <= 5, 'asfixia moderada',
             np.where(df['APGAR5'] <= 7, 'asfixia leve', 'normal')))

In [9]:
df['asfixia'].value_counts()

asfixia
normal              26566
asfixia leve          320
asfixia severa         74
asfixia moderada       68
Name: count, dtype: int64

In [10]:
# Caro(a) monitor(a), se puder comentar porque o código abaixo não funcionou como esperado eu agradeceria.

df.loc[df['APGAR5'] <= 3, 'asfixia'] = 'asfixia severa'
df.loc[(df['APGAR5'] >= 4) or (df['APGAR5'] <= 5), 'asfixia'] = 'asfixia moderada'
df.loc[(df['APGAR5'] >= 6) or (df['APGAR5'] <= 7), 'asfixia'] = 'asfixia leve'
df.loc[(df['APGAR5'] >= 8), 'asfixia'] = 'normal'


ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [11]:
# 9) Renomeie as variáveis para que fiquem no snake case, ou seja, em letras minúsculas, com um underscore entre as palávras. 

df_novo = df[['LOCNASC', 'IDADEMAE', 'ESTCIVMAE', 'ESCMAE', 'QTDFILVIVO', 
    'GESTACAO', 'GRAVIDEZ', 'CONSULTAS', 'APGAR5']]

novas_colunas = ['loc_nasc', 'idade_mae', 'est_civ_mae', 'esc_mae', 'qtd_fil_vivo', 
                 'gestacao', 'gravidez', 'consultas', 'agpar_5']

df_novo.columns = novas_colunas

df_novo

loc_nasc  idade_mae  est_civ_mae          esc_mae  qtd_fil_vivo  \
0             1         19          5.0      8 a 11 anos           0.0   
1             1         29          2.0      8 a 11 anos           1.0   
2             1         37          9.0      8 a 11 anos           2.0   
3             1         30          5.0  12 anos ou mais           0.0   
4             1         30          2.0      8 a 11 anos           1.0   
...         ...        ...          ...              ...           ...   
27023         1         32          2.0  12 anos ou mais           1.0   
27024         1         19          1.0      8 a 11 anos           0.0   
27025         1         24          2.0      8 a 11 anos           0.0   
27026         1         21          2.0      8 a 11 anos           1.0   
27027         1         18          5.0      8 a 11 anos           1.0   

              gestacao gravidez  consultas  agpar_5  
0      37 a 41 semanas    Única          4     10.0  
1      37 a 41 semanas    Única          4      9.0  
2      37 a 41 semanas    Única          4     10.0  
3      37 a 41 semanas    Única          3     10.0  
4      37 a 41 semanas    Única          4     10.0  
...                ...      ...        ...      ...  
27023  32 a 36 semanas    Única          4      9.0  
27024  37 a 41 semanas    Única          4      9.0  
27025  37 a 41 semanas    Única          4     10.0  
27026  32 a 36 semanas    Única          4      9.0  
27027  37 a 41 semanas    Única          4      8.0  

[27028 rows x 9 columns]